Don't prepend header in BDX messages

holbrookt opened the pull request:
#### Problem
What is being fixed?  Examples:
* `TransferSession` was prepending outgoing messages with a `PayloadHeader`, but this is not compatible with `ExchangeContext::SendMessage()` which creates its own `PayloadHeader` to prepend
* Fixes #8098

#### Change overview
- remove code in BdxTransferSession that prepended PayloadHeaders to
outgoing messages
- add MessageTypeData to convey message type and protocol
- add header prepending in TestBdxTransferSession.cpp
- fix some comments

#### Testing
Fixed `TestBdxTransferSession` to not expect messages to already have a `PayloadHeader`. Tests pass.

bzbarsky-apple commented on src/protocols/bdx/BdxTransferSession.h:
```diff
@@ -80,12 +80,22 @@ class DLL_EXPORT TransferSession
 
+    struct MessageTypeData
+    {
+        uint32_t ProtocolId = 0; // Should only be SecureChannel or BDX
```
Why not store this as an actual protocol id, as opposed to `uint32_t`?

holbrookt replied:
When I originally wrote this header, I ran into compilation errors when I tried using non-trivial types in the structs included in the anonymous union.

That being said, I tried doing this change today and got it to compile locally, so I'm not sure what I did wrong last time. I agree with you that it's much simpler and cleaner to use `Protocols::Id`

holbrookt replied:
New commit has these changes

bzbarsky-apple commented on src/protocols/bdx/tests/TestBdxTransferSession.cpp:
```diff
@@ -80,29 +80,44 @@ CHIP_ERROR ReadAndVerifyTLVString(nlTestSuite * inSuite, void * inContext, const
     {
-        NL_TEST_ASSERT(inSuite, false);
-        return;
+        payloadHeader.SetMessageType(Protocols::BDX::Id, typeData.MessageType);
```
If the typedata stored a Protocols::Id, could just have a single `SetMessageType` call here.

holbrookt replied:
Good catch, will implement

bzbarsky-apple commented on src/protocols/bdx/BdxTransferSession.h:
```diff
@@ -97,6 +107,7 @@ class DLL_EXPORT TransferSession
             TransferAcceptData transferAcceptData;
             BlockData blockdata;
             StatusReportData statusData;
+            MessageTypeData msgTypeData;
```
Is this used in practice anywhere outside the unit test?  I'm not obviously seeing it....

holbrookt replied:
It is - see `PrepareOutgoingMessageEvent()` in BdxTransferSession.cpp

Also this struct will be crucial to any application that uses `TransferSession` because it will need to pass the ProtocolId and MessageType to `ExchangeContext::SendMessage()`

bzbarsky-apple replied:
Ah, it's actually used as an arg to `AttachHeaderAndSend` looks like.  OK, that's the part I was missing.

holbrookt commented:
I added one more commit to fix HandleMessageReceived(). It was expecting the `PayloadHeader` to have not been consumed yet, but `ExchangeDelegate` does this already. I changed it to take the decoded `PayloadHeader` as an argument instead.
